### DISCUSSION TITLE
How to access properties in MVTLayer in pydeck for colors and tooltip?

### DISCUSSION THREAD

**Discussion by kopp** [visgl/deck.gl#8032] [Q&A]
I am having issues to access properties in vector tiles in pydeck.

My data looks like this:

```json
{
  "type": "FeatureCollection",
  "features": [
    {
      "type": "Feature",
      "id": 1,
      "geometry": {
        "type": "LineString",
        "coordinates": [
          [8.106825, 50.809013],
          // ...
          [8.101244, 50.816121]
        ]
      },
      "properties": { "color": [96, 177, 69], "text": "fancy" }
    },
    {
      "type": "Feature",
      "id": 2,
      "geometry": {
        "type": "LineString",
        "coordinates": [
```

I use tippecanoe to create vector tiles from that.  I would like to use the `color` property as color of the features.
I also want to display a tooltip that should use the `text` property.

My code using pydeck looks like this:

```python
    pydeck.Deck(
        layers=[
            pydeck.Layer(
                "MVTLayer",
                "/app/static/coverage/{z}/{x}/{y}.pbf",
                # get_line_color="properties.color",
                get_line_color='[properties.color[0], properties.color[1], properties.color[2]]',
                get_fill_color='[properties.color[0], properties.color[1], properties.color[2]]',  # I also tried 'properties.color', or 'color' or '{color}' or '{properties.color}'
                get_line_width=20,
                line_width_min_pixels=1,
                pickable=True,
                auto_highlight=True,
                highlight_color=[255, 255, 0],
            )
        ],
        tooltip={"text": "{text}"},
        map_style="light",
    )
```

I would now expect to see each feature with "its" color and a text hover that shows the content of the `text` property.

What I see instead is that the lines are of black/greyish color, all have the same, and that the tooltip is literally `{text}`.

Can you please point me to a working example that demonstrates how to access the properties in vector tiles?

**Comment by diehl** [comment]
@kopp Take a look at the following example that covers your tooltip question. 
https://colab.research.google.com/drive/1yx91WiXVe80a034J_o_6ThuYzKzO1aHP?usp=sharing
I think you are on the right track.

**Reply by diehl** [reply]
@kopp And here's another chunk of code that I have working that shows the use of properties to set the fill color. Does this help?
```
INITIAL_VIEW_STATE = pdk.ViewState(
  latitude=38,
  longitude=-97,
  zoom=3,
  min_zoom=3,
  max_zoom=16,
  pitch=0,
  bearing=0,
)

# Eliminates click state update delay
view = pdk.View(type="MapView", controller={"doubleClickZoom": False})

mvt_layer = pdk.Layer(
    "MVTLayer",
    "https://a.tiles.mapbox.com/v4/mapbox.mapbox-streets-v7/{z}/{x}/{y}.vector.pbf?access_token="+ACCESS_TOKEN,
    get_line_color = [192, 192, 192],
    get_fill_color = "properties.class == 'national_park' ? [255, 0, 0] : (properties.class == 'wood' ? [0, 255, 0] : [140, 170, 180])",
    line_width_min_pixels = 1,
    pickable=True
)

r = pdk.Deck(
    layers=[mvt_layer],
    initial_view_state=INITIAL_VIEW_STATE,
    views=[view],
    map_provider=None
)

map = pn.pane.DeckGL(r,sizing_mode='stretch_both')  
map.show()
```

**Reply by kopp** [reply]
Hi @diehl , thanks a lot already for the examples.  I guess what I'm missing is how to generate the vector tiles in the first place so that I get the properties like I need them for this to work.  I usually generate geojson files and from that vector tiles using tippecanoe -- how did you generate your tiles?

**Reply by kopp** [reply]
So I now have data that contains two properties, `heading` and `color`.

```{
  "type": "Feature",
  "id": 877,
  "geometry": {
    "type": "LineString",
    "coordinates": [
      [
        8.089311,
        49.847706
      ],
      // ...
    ]
  },
  "properties": {
    "heading": 0.533383366464606,
    "color": [
      187,
      222,
      107
    ]
  }
}

```

I can kind of make use of them:

```python
    TOOLTIP_HTML = ('<b>Heading: {properties.heading}</b><br />'
                '<b>heading retry</b>: {heading}<br />'
                '<b>colorb>: {color}<br />')

    pydeck.Deck(
        layers=[
            pydeck.Layer(
                "MVTLayer",
                "/app/static/coverage3/{z}/{x}/{y}.pbf",
                # working
                get_line_color='properties.heading < 0 ? [255, 0, 0] : [0, 0, 255]',
                get_fill_color='properties.heading < 0 ? [255, 0, 0] : [0, 0, 255]',
                # not working
                # get_line_color='properties.color',
                # get_fill_color='properties.color',
                get_line_width=20,
                line_width_min_pixels=1,
                pickable=True,
                auto_highlight=True,
                highlight_color=[255, 255, 0],
                # tooltip={"text": "{properties.color}"},
            )
        ],
        # not working (displays literal `{properties.heading}`)
        # tooltip={"text": "{properties.heading}"},
        # not working -- also displays literal {values}
        tooltip={'html': TOOLTIP_HTML},
        map_style="light",
    )
```

I wanted to use `JSON.parse` to parse the color (to make sure that it is interpreted as array), but that does not seem to be allowed.

Is there some way to "introspect" what's going on? (When I'm using pydeck in javascript, I can add a `console.log(properties)`, but here I cannot do that.)

Any further pointers would be greatly appreciated :)

What else could

**Reply by diehl** [reply]
@kopp Great question on the tile generation process. That is key. I follow the same workflow you do: (1) write geojson to disk (2) use tippecanoe to generate the tile. 

Here are the lessons I learned about the above workflow. The following is from my notes. 

The vector tiles must be derived from GeoJSON where a feature ID is present at the `feature.id` level as opposed to `feature.properties.id`. [Reference](https://github.com/visgl/deck.gl/discussions/6732#discussioncomment-2297421) When saving a GeoPandas geodataframe to disk as GeoJSON with `.to_file`, feature IDs will not be included and there is no option available to create `feature.id` fields. Exporting the GeoJSON in the following manner will create the feature IDs needed. 
```
with open('<GEOJSON FILENAME>', 'w') as f:
	f.write(gdf.to_json())
```
Additionally, Tippecanoe by default expects feature IDs that are numbers, not strings. The above operation will create feature IDs that are numbers represented by strings. Tippecanoe requires the addition of the `-aI` argument to convert the feature IDs from string representations of numbers to numbers when constructing the vector tiles. Auto highlight will work properly with the resulting vector tile representation of the data. [Reference](https://stackoverflow.com/questions/66366156/adding-a-feature-id-when-writing-to-file)

In terms of introspection, I have no wisdom to offer there. ;-) It's not easy to debug that is for sure!
